widgets/hwmontemp_linux.lua

```diff
@@ -25,30 +25,45 @@ local io = { open = io.open }
 local helpers = require"vicious.helpers"
 local spawn = require"vicious.spawn"
 
+local pathcache = {}
+
+local function read_sensor(path)
+    local f = io.open(path, "r")
+    local v = tonumber(f:read"*line") / 1000
+    f:close()
+
+    return v
+end
+
 -- hwmontemp: provides name-indexed temps from /sys/class/hwmon
 -- vicious.widgets.hwmontemp
 return helpers.setasyncall{
     async = function (format, warg, callback)
         if type(warg) ~= "table" or type(warg[1]) ~= "string" then
             return callback{}
         end
+
         local input = warg[2]
         if type(input) == "number" then
             input = ("temp%d_input"):format(input)
         else
             input = "temp1_input"
         end
 
-        spawn.easy_async_with_shell(
-            "grep " .. warg[1] .. " -wl /sys/class/hwmon/*/name",
-            function (stdout, stderr, exitreason, exitcode)
-                if exitreason == "exit" and exitcode == 0 then
-                    local f = io.open(stdout:gsub("name%s+", input), "r")
-                    callback{ tonumber(f:read"*line") / 1000 }
-                    f:close()
-                else
-                    callback{}
-                end
-            end)
+        local sensor = warg[1]
+        if pathcache[sensor] then
+            callback{ read_sensor(pathcache[sensor] .. input) }
+        else
+            spawn.easy_async_with_shell(
+                "grep " .. sensor .. " -wl /sys/class/hwmon/*/name",
+                function (stdout, stderr, exitreason, exitcode)
+                    if exitreason == "exit" and exitcode == 0 then
+                        pathcache[sensor] = stdout:gsub("name%s+", "")
+                        callback{ read_sensor(pathcache[sensor] .. input) }
+                    else
+                        callback{}
+                    end
+                end)
```

Comment:
Something like the following will not rerun grep if the given sensor is not found:

```suggestion
        if pathcache[sensor] == nil then
            spawn.easy_async_with_shell(
                "grep " .. sensor .. " -wl /sys/class/hwmon/*/name",
                function (stdout, stderr, exitreason, exitcode)
                    if exitreason == "exit" and exitcode == 0 then
                        pathcache[sensor] = stdout:gsub("name%s+", "")
                        callback{ read_sensor(pathcache[sensor] .. input) }
                    else
                        pathcache[sensor] = false
                        callback{}
                    end
                end)
        elseif pathcache[sensor] then
            callback{ read_sensor(pathcache[sensor] .. input) }
        else
            callback{}
```

Reply:
Re-running `grep` if the sensor could not be found was actually intended. I thought of the (admittedly rare) case of hot-pluggable devices that may occur after the window manager has been started.

I think it's a matter of taste. Invalid sensors should not be the average case but I get your point.

Reply:
Sounds fair to me.